### PR TITLE
Top stories

### PR DESCRIPTION
**Pull request by leggsimon**
This adds a new `/__offline/top-stories` route and controller.

*N.b.* Apologies for the large PR, ignore the files in `data` `helpers` `libs` `sections` which were all ported from front-page

- Copy over bulk of data loading and validation to fetch graph-ql data from front-page
- Setup n-ui for compilation and templates
- Disable sticky header
- Add custom styling for offline header
- Add custom data model for sub-navigation
- Refactor template partials
- Use vanilla layout

### Todo
- [x] Decorate response with `Link` headers of sub-resources for precaching
- [ ] Clean up data-store libraries copied over from front-page
- [ ] Clean up dependencies added to get n-ui to work

screenshot 2016-11-23 15 26 58

### CODE REVIEW

**Comment by quarterto** [comment]
@leggsimon could you possibly split the n-ui parts of this out into a separate PR?

**Comment by quarterto** [comment]
Rebased on latest master (n-ui) and history cleaned up

File: client/components/navigation/_main.scss [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+		display: block;
+	}
+}
+
+
+.o-header__nav--desktop {
+	@include oGridRespondTo(L) {
+		display: none;
+
+		.js-online {
```
Comment (leggsimon) [inline client/components/navigation/_main.scss]: missing `&`

File: server/controllers/top-stories.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
+const setNavigation = require('../libs/navigation');
+const dataStore = require('../data/data-store');
+const url = require('url')
+
+function imageSrcToImageServiceUrl (rawSrc) {
```
Comment (leggsimon) [inline server/controllers/top-stories.js]: move into helper

File: server/controllers/top-stories.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+		},
+		layout: 'vanilla',
+		withAssetPrecache: true
+	};
+
+	// Decorate response with link headers for subresources
+	const data = dataStore.retrieve('ukTopStories'); // FIXME: Use render params to get edition
+
+	data.topStories.items.forEach(content => {
+		const contentPath = `/__offline${url.parse(content.url).path}`
```
Comment (leggsimon) [inline server/controllers/top-stories.js]: change path to `/offline`

File: server/controllers/top-stories.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+		res.linkResource(contentPath, {as: 'document', rel: 'precache'})
+
+		if (content.primaryImage && content.primaryImage.rawSrc) {
+			const imagePath = imageSrcToImageServiceUrl(content.primaryImage.rawSrc);
+			res.linkResource(imagePath, {as: 'image', rel: 'precache'})
+		}
+	});
+
+	res.set('Surrogate-Control', 'max-age=30, stale-while-revalidate=60, stale-if-error=86400');
```
Comment (leggsimon) [inline server/controllers/top-stories.js]: max age is very low

File: server/controllers/top-stories.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	res.set('Cache-Control', res.FT_NO_CACHE);
+	res.set('Access-Control-Allow-Origin', '*');
+
+	const template = 'top-stories';
+
+	res.locals = setNavigation(res.locals, 'Home');
+	res.locals.navigationLists.navbar_mobile = [
+		{ name: 'Top stories', href: '' },
```
Comment (leggsimon) [inline server/controllers/top-stories.js]: add `hrefs`

Comment (leggsimon) [inline server/controllers/top-stories.js]: lose this object

File: server/libs/navigation.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
+const data = [
+	{ name: 'Home', href: '/__offline/top-stories' },
+	{ name: 'Following', href: '/__offline/following' },
+	{ name: 'Saved articles', href: '/__offline/saved-articles' },
+];
```
Comment (leggsimon) [inline server/libs/navigation.js]: Update labels

**Pull request merged**